// rosaenlg/dist/LanguageOther.js
"use strict";
/**
 * @license
 * Copyright 2019 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageOther = void 0;
const LanguageImpl_1 = require("./LanguageImpl");
const locale_1 = require("date-fns/locale");
class LanguageOther extends LanguageImpl_1.LanguageImpl {
    constructor(languageCommon) {
        super(languageCommon);
        // let's be very permissive
        this.hasGender = false; // as we have no dict to check
        this.hasNeutral = true; // why not!
        this.langForDateFns = locale_1.enUS;
        this.spacesWhenSeparatingElements = true;
        this.iso2 = languageCommon.iso2;
    }
    // well this is a strange idea
    getAgreeAdj(adjective /* , _gender: Genders, _number: Numbers, _subject: any, _params: AgreeAdjParams */) {
        return adjective;
    }
    getSubstantive(subst, number) {
        if (number == 'S') {
            return subst;
        }
        else {
            const err = new Error();
            err.name = 'InvalidArgumentError';
            err.message = `cannot make plural of word ${subst} in ${this.iso2}`;
            throw err;
        }
    }
}
exports.LanguageOther = LanguageOther;
