// rosaenlg/dist/RandomManager.js
"use strict";
/**
 * @license
 * Copyright 2019 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RandomManager = void 0;
const random_js_1 = require("random-js");
class RandomManager {
    constructor(randomSeed) {
        this.incrRandomer = 10;
        this.rndNextPos = 0;
        this.rndTable = [];
        this.rndEngine = new random_js_1.Random(random_js_1.MersenneTwister19937.seed(randomSeed));
    }
    getRndNextPos() {
        return this.rndNextPos;
    }
    setRndNextPos(rndNextPos) {
        this.rndNextPos = rndNextPos;
    }
    /*
    public setSpy(spy: Spy): void {
      this.spy = spy;
    }
    */
    incrRnd(incr) {
        for (let i = 0; i < incr; i++) {
            this.getNextRnd(); // we don't care about the result
        }
    }
    getNextRnd() {
        if (this.rndNextPos >= this.rndTable.length) {
            for (let i = 0; i < this.incrRandomer; i++) {
                // has biaises: https://www.npmjs.com/package/random-js; find better?
                this.rndTable.push(this.rndEngine.real(0, 1, false));
            }
        }
        return this.rndTable[this.rndNextPos++];
    }
    getItemWeight(params, item) {
        return (params[item] && params[item].weight) || 1;
    }
    getSumOfWeights(max, params) {
        let sumOfWeights = 0;
        for (let i = 1; i <= max; i++) {
            sumOfWeights += this.getItemWeight(params, i);
        }
        return sumOfWeights;
    }
    /*
      https://stackoverflow.com/questions/6443176/how-can-i-generate-a-random-number-within-a-range-but-exclude-some
      https://medium.com/@peterkellyonline/weighted-random-selection-3ff222917eb6
    */
    getTargetIndex(origIndex, excludes) {
        let targetIndex = 0;
        for (let i = 1; i <= origIndex; i++) {
            targetIndex++;
            while (excludes.indexOf(targetIndex) > -1) {
                targetIndex++;
            }
        }
        return targetIndex;
    }
    getWeightedRandom(max, weights) {
        const sumOfWeights = this.getSumOfWeights(max, weights);
        let randomWeight = Math.floor(this.getNextRnd() * sumOfWeights) + 1;
        for (let i = 1; i <= max; i++) {
            randomWeight = randomWeight - this.getItemWeight(weights, i);
            if (randomWeight <= 0) {
                return i;
            }
        }
    }
    randomNotIn(max, weights, excludes) {
        if (excludes.length === max) {
            // it won't be possible to find a new one
            return null;
        }
        //il faut translater les index des poids
        const translatedWeights = {};
        let newIndex = 0;
        for (let i = 1; i <= max; i++) {
            if (excludes.indexOf(i) === -1) {
                newIndex++;
                translatedWeights[newIndex] = { weight: this.getItemWeight(weights, i) };
            }
        }
        const weightedRandom = this.getWeightedRandom(max - excludes.length, translatedWeights);
        // inverse mapping
        return this.getTargetIndex(weightedRandom, excludes);
    }
}
exports.RandomManager = RandomManager;
