// synonym-optimizer/dist/LanguageSyn.js
"use strict";
/**
 * @license
 * Copyright 2019 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageSyn = void 0;
const tokenizer = require("wink-tokenizer");
const rosaenlg_filter_1 = require("rosaenlg-filter");
class LanguageSyn {
    extractWords(input) {
        // console.log(`tokenizing: ${input}`);
        const myTokenizer = new tokenizer();
        myTokenizer.defineConfig({
            currency: false,
            number: false,
            punctuation: false,
            symbol: false,
            time: false,
        });
        const tokenized = myTokenizer.tokenize(input);
        // console.log(`tokenized: ${tokenized}`);
        const res = [];
        tokenized.forEach((elt) => {
            // no alien tags and no html elements
            if (elt.tag != 'alien' &&
                rosaenlg_filter_1.blockLevelHtmlElts.indexOf(elt.value) == -1 &&
                rosaenlg_filter_1.inlineHtmlElts.indexOf(elt.value) == -1) {
                res.push(elt.value);
            }
        });
        // console.log(`res: ${res}`);
        return res;
    }
}
exports.LanguageSyn = LanguageSyn;
