// rosaenlg-commons/dist/index.js
"use strict";
/**
 * @license
 * Copyright 2019 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIso2fromLocale = exports.buildLanguageCommon = exports.Constants = exports.DictManager = exports.LanguageCommonOther = exports.LanguageCommonSpanish = exports.LanguageCommonItalian = exports.LanguageCommonGerman = exports.LanguageCommonFrench = exports.LanguageCommonEnglish = exports.LanguageCommon = void 0;
var LanguageCommon_1 = require("./LanguageCommon"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommon", { enumerable: true, get: function () { return LanguageCommon_1.LanguageCommon; } });
var LanguageCommonEnglish_1 = require("./LanguageCommonEnglish"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommonEnglish", { enumerable: true, get: function () { return LanguageCommonEnglish_1.LanguageCommonEnglish; } });
var LanguageCommonFrench_1 = require("./LanguageCommonFrench"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommonFrench", { enumerable: true, get: function () { return LanguageCommonFrench_1.LanguageCommonFrench; } });
var LanguageCommonGerman_1 = require("./LanguageCommonGerman"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommonGerman", { enumerable: true, get: function () { return LanguageCommonGerman_1.LanguageCommonGerman; } });
var LanguageCommonItalian_1 = require("./LanguageCommonItalian"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommonItalian", { enumerable: true, get: function () { return LanguageCommonItalian_1.LanguageCommonItalian; } });
var LanguageCommonSpanish_1 = require("./LanguageCommonSpanish"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommonSpanish", { enumerable: true, get: function () { return LanguageCommonSpanish_1.LanguageCommonSpanish; } });
var LanguageCommonOther_1 = require("./LanguageCommonOther"); /* istanbul ignore next */
Object.defineProperty(exports, "LanguageCommonOther", { enumerable: true, get: function () { return LanguageCommonOther_1.LanguageCommonOther; } });
var DictManager_1 = require("./DictManager");
Object.defineProperty(exports, "DictManager", { enumerable: true, get: function () { return DictManager_1.DictManager; } });
var Constants_1 = require("./Constants");
Object.defineProperty(exports, "Constants", { enumerable: true, get: function () { return Constants_1.Constants; } });
var helper_1 = require("./helper");
Object.defineProperty(exports, "buildLanguageCommon", { enumerable: true, get: function () { return helper_1.buildLanguageCommon; } });
Object.defineProperty(exports, "getIso2fromLocale", { enumerable: true, get: function () { return helper_1.getIso2fromLocale; } });
