// rosaenlg-pug-code-gen/dist/LanguageCodeGenEnglish.js
"use strict";
/**
 * @license
 * Copyright 2018, Ludan Stoecklé
 * SPDX-License-Identifier: MIT
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageCodeGenEnglish = void 0;
const LanguageCodeGen_1 = require("./LanguageCodeGen");
const english_plurals_1 = require("english-plurals");
const plurals_json_1 = __importDefault(require("english-plurals-list/dist/plurals.json"));
const verbs_json_1 = __importDefault(require("english-verbs-irregular/dist/verbs.json"));
const gerunds_json_1 = __importDefault(require("english-verbs-gerunds/dist/gerunds.json"));
const english_verbs_helper_1 = require("english-verbs-helper");
class LanguageCodeGenEnglish extends LanguageCodeGen_1.LanguageCodeGen {
    constructor() {
        super();
        this.iso2 = 'en';
        this.hasFlexVerbs = true;
        this.hasFlexWords = true; // is meaningless for setRefGender, but useful for plurals, in value
        this.hasFlexAdjectives = false;
        this.mergedVerbsDataEn = english_verbs_helper_1.mergeVerbsData(verbs_json_1.default, gerunds_json_1.default);
    }
    getVerbInfo(verb) {
        return english_verbs_helper_1.getVerbInfo(this.mergedVerbsDataEn, verb);
    }
    getWordInfo(word) {
        return { plural: english_plurals_1.getPlural(null, plurals_json_1.default, word) };
    }
}
exports.LanguageCodeGenEnglish = LanguageCodeGenEnglish;
