// rosaenlg-pug-code-gen/dist/LanguageCodeGenSpanish.js
"use strict";
/**
 * @license
 * Copyright 2018, Ludan Stoecklé
 * SPDX-License-Identifier: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageCodeGenSpanish = void 0;
const LanguageCodeGen_1 = require("./LanguageCodeGen");
const spanish_adjectives_wrapper_1 = require("spanish-adjectives-wrapper");
const spanish_words_1 = require("spanish-words");
const spanish_verbs_wrapper_1 = require("spanish-verbs-wrapper");
class LanguageCodeGenSpanish extends LanguageCodeGen_1.LanguageCodeGen {
    constructor() {
        super(...arguments);
        this.iso2 = 'es';
        this.hasFlexVerbs = true;
        this.hasFlexWords = true;
        this.hasFlexAdjectives = true;
    }
    getVerbInfo(verb) {
        return spanish_verbs_wrapper_1.getVerbInfo(verb);
    }
    getWordInfo(word) {
        return spanish_words_1.getWordInfo(word);
    }
    getAdjectiveInfo(adjective) {
        return spanish_adjectives_wrapper_1.getAdjectiveInfo(adjective);
    }
}
exports.LanguageCodeGenSpanish = LanguageCodeGenSpanish;
