// rosaenlg-pug-code-gen/dist/LanguageCodeGenItalian.js
"use strict";
/**
 * @license
 * Copyright 2018, Ludan Stoecklé
 * SPDX-License-Identifier: MIT
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageCodeGenItalian = void 0;
const LanguageCodeGen_1 = require("./LanguageCodeGen");
const italian_adjectives_1 = require("italian-adjectives");
const adjectives_json_1 = __importDefault(require("italian-adjectives-dict/dist/adjectives.json"));
const italian_words_1 = require("italian-words");
const words_json_1 = __importDefault(require("italian-words-dict/dist/words.json"));
const italian_verbs_1 = require("italian-verbs");
const verbs_json_1 = __importDefault(require("italian-verbs-dict/dist/verbs.json"));
class LanguageCodeGenItalian extends LanguageCodeGen_1.LanguageCodeGen {
    constructor() {
        super(...arguments);
        this.iso2 = 'it';
        this.hasFlexVerbs = true;
        this.hasFlexWords = true;
        this.hasFlexAdjectives = true;
    }
    getVerbInfo(verb) {
        return italian_verbs_1.getVerbInfo(verbs_json_1.default, verb);
    }
    getWordInfo(word) {
        return italian_words_1.getWordInfo(words_json_1.default, word);
    }
    getAdjectiveInfo(adjective) {
        return italian_adjectives_1.getAdjectiveInfo(adjectives_json_1.default, adjective);
    }
}
exports.LanguageCodeGenItalian = LanguageCodeGenItalian;
