// rosaenlg-pug-code-gen/dist/LanguageCodeGen.js
"use strict";
/**
 * @license
 * Copyright 2018, Ludan Stoecklé
 * SPDX-License-Identifier: MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageCodeGen = void 0;
class LanguageCodeGen {
    constructor() {
        // do nothing
    }
    // more of a helper
    getVerbsInfo(verbs) {
        const res = {};
        for (const verb of verbs) {
            res[verb] = this.getVerbInfo(verb);
        }
        return res;
    }
    getWordsInfo(words) {
        const res = {};
        for (const word of words) {
            res[word] = this.getWordInfo(word);
        }
        return res;
    }
    getAdjectivesInfo(adjectives) {
        const res = {};
        for (const adjective of adjectives) {
            res[adjective] = this.getAdjectiveInfo(adjective);
        }
        return res;
    }
    // istanbul ignore next
    getVerbInfo(_verb) {
        const err = new Error();
        err.name = 'InvalidArgumentException';
        err.message = `cannot getVerbInfo in ${this.iso2}`;
        throw err;
    }
    // istanbul ignore next
    getWordInfo(_word) {
        const err = new Error();
        err.name = 'InvalidArgumentException';
        err.message = `cannot getWordInfo in ${this.iso2}`;
        throw err;
    }
    // istanbul ignore next
    getAdjectiveInfo(_adjective) {
        const err = new Error();
        err.name = 'InvalidArgumentException';
        err.message = `cannot getAdjectiveInfo in ${this.iso2}`;
        throw err;
    }
}
exports.LanguageCodeGen = LanguageCodeGen;
