// rosaenlg-pug-code-gen/dist/LanguageCodeGenGerman.js
"use strict";
/**
 * @license
 * Copyright 2018, Ludan Stoecklé
 * SPDX-License-Identifier: MIT
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageCodeGenGerman = void 0;
const LanguageCodeGen_1 = require("./LanguageCodeGen");
const words_json_1 = __importDefault(require("german-words-dict/dist/words.json"));
const german_words_1 = require("german-words");
const german_adjectives_1 = require("german-adjectives");
const adjectives_json_1 = __importDefault(require("german-adjectives-dict/dist/adjectives.json"));
const german_verbs_1 = require("german-verbs");
const verbs_json_1 = __importDefault(require("german-verbs-dict/dist/verbs.json"));
class LanguageCodeGenGerman extends LanguageCodeGen_1.LanguageCodeGen {
    constructor() {
        super(...arguments);
        this.iso2 = 'de';
        this.hasFlexVerbs = true;
        this.hasFlexWords = true;
        this.hasFlexAdjectives = true;
    }
    getVerbInfo(verb) {
        return german_verbs_1.getVerbInfo(verbs_json_1.default, verb);
    }
    getWordInfo(word) {
        return german_words_1.getWordInfo(words_json_1.default, word);
    }
    getAdjectiveInfo(adjective) {
        return german_adjectives_1.getAdjectiveInfo(adjectives_json_1.default, adjective);
    }
}
exports.LanguageCodeGenGerman = LanguageCodeGenGerman;
